---
title: "RNA-Seq"
author: "Mojca Juter≈°ek"
date: "`r Sys.Date()`"
output:
  html_document:
    fig_caption: yes
    self_contained: yes
    fig_width: 12
    fig_height: 9
    toc: yes
    toc_depth: 2
    toc_float: true
    number_sections: yes
    theme: flatly
    highlight: tango
  pdf_document:
    toc: yes
    toc_depth: '2'
  word_document:
    toc: yes
    toc_depth: '2'
editor_options:
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = c("png"), 
                      fig.align = "center", 
                      fig.height = 8, 
                      fig.width = 9,
                      fig.show = "hold",
                      results = "markup",
                      warning = FALSE, message = FALSE
)

rm(list = ls(all = TRUE))
set.seed(123456)
```

```{r,include=FALSE}
library("limma")
library("edgeR")
library("RColorBrewer")
library("ggplot2")
library("openxlsx")
library("data.table")
library("pheatmap")
```

# Data import

## Import counts

```{r}
fp <- file.path("..","input")
fn <- "featureCounts_all_multimap_counts_clean.txt"
counts <- read.table(file.path(fp,fn), header=TRUE, sep="\t")
dim(counts)

colnames(counts) <- gsub("^\\.\\.output\\.", "", colnames(counts))
colnames(counts) <- gsub("\\.Aligned\\.sortedByCoord\\.out\\.bam$", "", colnames(counts))
colnames(counts) <- sub("^(([^_]+)_([^_]+)_([^_]+))_.*", "\\1", colnames(counts))

rownames(counts) <- counts$GeneID
counts <- counts[,-1]
```

## Import phenodata

```{r}
fp <- file.path("..","..")
fn <- "phenodata.txt"
pheno <- read.table(file.path(fp,fn), header=TRUE, sep="\t")
```

```{r}
ind <- match(pheno$sampleID, colnames(counts))
all(pheno$sampleID == colnames(counts)[ind])
counts <- counts[,ind]
all(pheno$sampleID == colnames(counts))
```

# Define experimental groups

```{r}
group <- factor(pheno$Biological.group, levels = unique(pheno$Biological.group))
levels(group)
```

```{r}
## Create a DGEList object 
data <- edgeR::DGEList(counts = counts, group = group)
data$samples

```

```{r}
# Define the colours of groups for plotting
genotype_colors <- c(
  "FLA" = "#a6cee3",
  "SW5" = "#1f78b4",
  "SW5-7" = "#b2df8a",
  "SW7" = "#33a02c")
infection_colors <- c(
  "H" = "#66c2a5",
  "SNI" = "#fc8d62",
  "SRB" = "#8da0cb"
)
pheno$temperature <- factor(pheno$temperature, levels = c(22, 32))
temperature_colors <- c(
  "22" = "blue",
  "32" = "red"
)

genotype_col <- genotype_colors[pheno$Genotype]
pheno$infection <- trimws(pheno$infection, which = "right")
infection_col <- infection_colors[pheno$infection]
temperature_col <- temperature_colors[pheno$temperature]

genotypes <- levels(as.factor(pheno$Genotype))
infection <- levels(as.factor(pheno$infection))
temperature <- levels(as.factor(pheno$temperature))
```

# Filtering counts

```{r}
keep <- filterByExpr(data, group=group, min.count=100, min.total.count=200)
data_f <- data[keep, ,keep.lib.sizes=TRUE]
dim(data$counts)
dim(data_f$counts)

```

```{r}
opar <- par()
par(mfrow=c(1,2), cex = 0.6)

nsamples <- ncol(counts)


lcpm <- log(as.matrix(data$counts+1),2)
plot(density(lcpm), col="black", lwd=2, ylim=c(0,0.8), las=2, main="", xlab="")
title(main="A. Raw STAR counts", xlab="log2(counts)")
abline(v=0, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=genotype_col[i], lwd=2)
}

lcpm <- log(as.matrix(data_f$counts+1),2)
plot(density(lcpm), col="black", lwd=2, ylim=c(0,0.2), las=2, main="", xlab="")
title(main="B. Filtered STAR counts", xlab="log2(counts)")
abline(v=0, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=genotype_col[i], lwd=2)
}

par(opar)
```

# Calculate TMM normalisation factors

```{r}
data_f <- calcNormFactors(data_f)
```

```{r}
opar <- par()
par(mfrow=c(1,2), cex = 0.6)

nsamples <- ncol(counts)


lcpm <- log(as.matrix(data_f$counts+1),2)
plot(density(lcpm), col="black", lwd=2, ylim=c(0,0.2), las=2, main="", xlab="")
title(main="B. Filtered STAR counts", xlab="log2(counts)")
abline(v=0, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=genotype_col[i], lwd=2)
}


lcpm <- cpm(data_f$counts, log=TRUE)
plot(density(lcpm), col="black", lwd=2, ylim=c(0,0.21), las=2, main="", xlab="")
title(main="A. Normalised STAR counts", xlab="log2(counts)")
abline(v=0, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=genotype_col[i], lwd=2)
}

par(opar)
```

```{r}
opar <- par()
par(mfrow=c(1,3), cex = 0.6)
boxplot(log(data$counts+1,2), las=2, ylab="log2(counts)", col=genotype_col, main="Before filtering")
boxplot(log(data_f$counts+1,2), las=2, ylab="log2(counts)", col=genotype_col, main="After filterinf")
boxplot(cpm(data_f$counts, log=TRUE), ylab="log2(counts)", col=genotype_col, las=2, main="After normalisation")
par(opar)
```

# PCA plots

```{r}
logcpm = cpm(data_f, normalized.lib.sizes = TRUE, log = TRUE, prior.count = 1)
pca <- prcomp(t(logcpm))
```

```{r}
plot(pca$x[, 1:2], pch=20, cex=3, main="PCA")
```

```{r}
plot(pca$x[, 1:2], col=genotype_col, pch=20, cex=3, main="PCA by genotype")
legend("topright", legend=genotypes, bty = "n", pch = 20, col=genotype_colors, pt.cex = 3)
```

```{r}
plot(pca$x[, 1:2], col=infection_col, pch=20, cex=3, main="PCA by infection type")
legend("topright", legend=infection, bty = "n", pch = 20, col=infection_colors, pt.cex = 3)
```

```{r}
plot(pca$x[, 1:2], col=temperature_col, pch=20, cex=3, main="PCA by temperature")
legend("topright", legend=temperature, bty = "n", pch = 20, col=temperature_colors, pt.cex = 3)
```

```{r}
limma::plotMDS(logcpm,
               col = genotype_col,
               pch = 16,
               cex = 2,
               main = "MDS plot by genotype")
legend("topright",
       legend = genotypes,
       bty = "n",
       pch = 16,
       col = genotype_colors,
       ncol = 2,
       pt.cex = 2)
```

```{r}
limma::plotMDS(logcpm,
               col = genotype_col,
               labels=colnames(data_f),
               pch = 16,
               cex = 0.9,
               main = "MDS plot by genotype")
legend("topright",
       legend = genotypes,
       bty = "n",
       pch = 16,
       col = genotype_colors,
       ncol = 2,
       pt.cex = 2)
```

# DE analysis

```{r}
design <- model.matrix(~0+group)
colnames(design) <- levels(group)
rownames(design) <- pheno$sampleID
rownames(design)
```

```{r}
fit <- edgeR::voomLmFit(data_f, design, plot= TRUE)
```
